src/render_dataset_with_caching.py

```diff
@@ -0,0 +1,180 @@
+                        "render_params": render_params, "render_ids": render_ids})
+
+        with open(f"{self.cache_dir}/dataset.json", "w") as fout:
+            json.dump(dataset, fout)
+
+    def __len__(self):
+        return self.num_files
+
+    def __getitem__(self, idx):
+        # load some input_data for our render_function
+        im_2d_cube_ids = []
+        images = []
+
+        if self.cache_setting == "load":
+            dataset_entry = self.dataset[idx]
+            image = io.imread(dataset_entry["data_file"])
+            render_params = dataset_entry["render_params"]
+            render_ids = dataset_entry["render_ids"]
+            renders = dataset_entry["renders"]
+
+            # loop to generate a set of images of length: len(render_ids) *  camera_samples
+            for i, path in enumerate(renders):
+                # generate a rendered image of the given style
+                renderedimage = io.imread(path)
+                final_image = transforms.functional.to_tensor(renderedimage)
+                images.append(final_image)
+                im_2d_cube_ids.append(
+                    idx + render_ids[i//int(self.camera_samples)])
+
+        else: 
+            img_name = os.path.join(self.data_dir, self.all_files[idx])
+            image = io.imread(img_name)
+            render_params = []
+            render_ids = [i for i in range(self.num_psis_per_data_cube)]
+            images_names = []
+
+            # prepare the sampler of camera transforms
+            camera_degree_range = 45.0
+            apply_camera = transforms.RandomRotation(
+                camera_degree_range, resample=PIL.Image.BICUBIC
+            )
+
+            # loop to generate a set of images with different styles and different camera angles
+            for k in range(self.num_psis_per_data_cube):
+                # generate a repeatable set of render parameters for our render_function
+                random.seed(a=(idx + k))
+                convfilter = [random.random() for i in range(9)]
+
+                for j in range(self.camera_samples):
+                    # generate a rendered image of the given style
+                    renderedimage = render_function(image.transpose(2, 1, 0), convfilter, apply_camera)
+                    final_image = transforms.functional.to_tensor(renderedimage)
+                    images.append(final_image)
+                    im_2d_cube_ids.append(idx + render_ids[j//int(self.camera_samples)])
+                    render_params.append(convfilter)
+                    
+                    if self.cache_setting == "save":
+                        outpath = f"{self.cache_dir}/{self.all_files[idx]}_rendered_{j}_{k}.png"
+                        images_names.append(outpath)
+                        renderedimage.save(outpath)
+
+
+            if self.cache_setting == "save":
+                self.save_to_dataset_file(img_name, images_names, render_params, render_ids)
```

Comment:
Now that I'm seeing this code, for repeatability, save_to_dataset_file needs to make sure it inserts at `idx`, in case `__getitem__` is not called sequentially.  Which screws up being able to store this as an array, unless you mandate that caching has to happen sequentially.

Reply:
And I actually *would* mandate that caching has to happen sequentially. :)